File: data/code_pairs/fortran/500.f90
program test_q
    implicit none

contains

    subroutine qr(f)
        interface
            subroutine f(x)
                real, intent(in) :: x
            end subroutine f
        end interface
        external f

        call f(5.0)
    end subroutine qr

    subroutine qc(f)
        interface
            subroutine f(x)
                complex, intent(in) :: x
            end subroutine f
        end interface
        external f

        call f((5.0, 0.0))
    end subroutine qc

    subroutine realFunction(x)
        real, intent(in) :: x
        print *, 'Real function called with argument:', x
    end subroutine realFunction

    subroutine complexFunction(x)
        complex, intent(in) :: x
        print *, 'Complex function called with argument:', x
    end subroutine complexFunction

end program test_q

! To test, call qr(realFunction) and qc(complexFunction) within the program scope.

File: data/code_pairs/cpp/500.cpp
#include <iostream>
#include <functional>
#include <complex>

template<typename Real>
void qr(std::function<void(Real)> f) {
    f(42); // Call f with a real number
}

template<typename Complex>
void qc(std::function<void(Complex)> f) {
    f(std::complex<double>(42, 0)); // Call f with a complex number
}

void realFunction(double x) {
    std::cout << "Real function called with argument: " << x << std::endl;
}

void complexFunction(std::complex<double> z) {
    std::cout << "Complex function called with argument: " << z << std::endl;
}

int main() {
    qr<double>(realFunction);
    qc<std::complex<double>>(complexFunction);
    
    return 0;
}